### PR TITLE
Curio support

### PR DESCRIPTION
**Pull request by Darkere**
see #3012

### CODE REVIEW

**Comment by starforcraft** [comment]
lol https://www.curseforge.com/minecraft/mc-mods/refined-curios this mod is useless now

**Comment by raoulvdberge** [comment]
I don't believe such functionality needs to be an addon

**Comment by starforcraft** [comment]
> I don't believe such functionality needs to be an addon

yeah true was actually a bit stupid to do a addon mod instead of just a PR

**Comment by Darkere** [comment]
Lol. Did not even think to look if there is an addon. Timing is too good too :D

looks like the plan is to not merge this for 1.9.13 so you might still have some time before this happens.

File: src/main/java/com/refinedmods/refinedstorage/network/OpenNetworkItemMessage.java [hunk 1]
```diff
@@ -4,37 +4,62 @@
...
 
-    public OpenNetworkItemMessage(int slotId) {
+    public OpenNetworkItemMessage(int slotId, String curioSlot) {
         this.slotId = slotId;
+        this.curioSlot = curioSlot;
     }
 
     public static OpenNetworkItemMessage decode(PacketBuffer buf) {
-        return new OpenNetworkItemMessage(buf.readInt());
+        return new OpenNetworkItemMessage(buf.readInt(), buf.readString(32767));
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/network/OpenNetworkItemMessage.java:30]: I think we have a util function for this?

File: src/main/java/com/refinedmods/refinedstorage/network/OpenNetworkItemMessage.java [hunk 2]
```diff
@@ -4,37 +4,62 @@
...
         ServerPlayerEntity player = ctx.get().getSender();
 
         if (player != null) {
             ctx.get().enqueueWork(() -> {
-                ItemStack stack = player.inventory.getStackInSlot(message.slotId);
+                ItemStack stack = null;
+                if (message.curioSlot.isEmpty()) {
+                    stack = player.inventory.getStackInSlot(message.slotId);
+                } else {
+                    LazyOptional<ICuriosItemHandler> curiosHandler = CuriosApi.getCuriosHelper().getCuriosHandler(player);
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/network/OpenNetworkItemMessage.java:47]: Please split this code a bit more up, it's too much lines / too much indentation in a single method.

File: src/main/java/com/refinedmods/refinedstorage/screen/KeyInputListener.java [hunk 3]
```diff
@@ -3,63 +3,90 @@
...
-                    onError.accept(new TranslationTextComponent("misc.refinedstorage.network_item.shortcut_duplicate", new TranslationTextComponent(items[0].getTranslationKey())));
+                    sendError(new TranslationTextComponent("misc.refinedstorage.network_item.shortcut_duplicate", new TranslationTextComponent(items[0].getTranslationKey())));
                     return;
                 }
 
                 slotFound = i;
             }
         }
+        if (CuriosIntegration.isLoaded() && slotFound == -1) {
+            AtomicReference<Boolean> found = new AtomicReference<>();
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/KeyInputListener.java:59]: I think we can do better than this.

File: src/main/java/com/refinedmods/refinedstorage/screen/KeyInputListener.java [hunk 4]
```diff
@@ -3,63 +3,90 @@
...
                 if (slotFound != -1) {
-                    onError.accept(new TranslationTextComponent("misc.refinedstorage.network_item.shortcut_duplicate", new TranslationTextComponent(items[0].getTranslationKey())));
+                    sendError(new TranslationTextComponent("misc.refinedstorage.network_item.shortcut_duplicate", new TranslationTextComponent(items[0].getTranslationKey())));
                     return;
                 }
 
                 slotFound = i;
             }
         }
+        if (CuriosIntegration.isLoaded() && slotFound == -1) {
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/KeyInputListener.java:58]: Please split this method a bit more up.

**Review by raoulvdberge — CHANGES_REQUESTED** [review]

**Review by raoulvdberge — COMMENTED** [review]
LGTM. Did you try launching the game without Curios installed to see if it won't break?

**Comment by Darkere** [comment]
Working just fine.

**Comment by raoulvdberge** [comment]
Thx

**Pull request merged**